e2e/main.sh

```diff
@@ -0,0 +1,197 @@
+#!/usr/bin/env bash
+
+set -euo pipefail
+
+SCRIPT_DIR="$(readlink -f "$(dirname "${BASH_SOURCE[0]}")")"
+OUTPUT_DIR="${ARTIFACT_DIR:-_output}/microshift-e2e-$(date +'%Y%m%d-%H%M%S')/"
+
+usage() {
+    echo "Usage: $(basename "${0}") {list|run} [filter]"
+    echo ""
+    echo "   list      Lists tests"
+    echo "   run       Runs tests"
+    echo "   filter    Simple string to match against test files, e.g. 'reboot', 'boot', 'smoke'"
+    echo ""
+    echo " Script expects two environmental variables:"
+    echo "   - USHIFT_IP"
+    echo "   - USHIFT_USER"
+    echo ""
+    echo " Script assumes following:"
+    echo "   - Passwordless SSH to \$USHIFT_USER@\$USHIFT_IP is configured"
+    echo "   - Both hosts already exchanged their public keys:"
+    echo "     - Test runner has MicroShift's sshd key in ~/.ssh/known_keys"
+    echo "     - Remote \$USHIFT_USER has test runner's key in ~/.ssh/authorized_keys"
+    echo "   - Passwordless sudo for \$USHIFT_USER"
+    echo ""
+    echo " Script aims to be target platform agnostic. It means that for some environments (e.g. GCP)"
+    echo " it might be required to export firewall::open_port and firewall::close_port functions"
+    echo " so the tests can open custom ports"
+
+    [ -n "$1" ] && echo -e "\nERROR: $1"
+    exit 1
+}
+
+log() {
+    echo -e "$(date +'%H:%M:%S.%N')    $*"
+}
+
+var_should_not_be_empty() {
+    local var_name=${1}
+    if [ -z "${!var_name+x}" ]; then
+        echo >&2 "Environmental variable '${var_name}' is unset"
+        return 1
+    elif [ -z "${!var_name}" ]; then
+        echo >&2 "Environmental variable '${var_name}' is empty"
+        return 1
+    fi
+}
+
+function_should_be_exported() {
+    local fname=${1}
+    if ! declare -F "${fname}"; then
+        log "WARNING: Function '${fname}' is unexported. It is expected that function is provided for interacting with cloud provider"
+        return 1
+    fi
+}
+
+check_passwordless_ssh() {
+    ssh -o BatchMode=yes "${USHIFT_USER}@${USHIFT_IP}" "true" || {
```

Comment:
There are quite a few repetitions for the `ssh` command invocation. If we create a `run_ssh_command` function, it should also allow us to transparently execute commands locally if the host/user are the same.